packages/firestore/test/integration/api/bundle.test.ts

```diff
@@ -0,0 +1,257 @@
+  expect(p.bytesLoaded).to.be.equal(p.totalBytes);
+  expect(p.documentsLoaded).to.equal(p.totalDocuments);
+}
+
+function verifyInProgress(
+  p: firestore.LoadBundleTaskProgress,
+  expectedDocuments: number
+): void {
+  expect(p.taskState).to.equal('Running');
+  expect(p.bytesLoaded).lte(p.totalBytes);
+  expect(p.documentsLoaded).lte(p.totalDocuments);
+  expect(p.documentsLoaded).to.equal(expectedDocuments);
+}
+
+apiDescribe('Bundles', (persistence: boolean) => {
+  const encoder = new TextEncoder();
+  const testDocs: { [key: string]: firestore.DocumentData } = {
+    a: { k: { stringValue: 'a' }, bar: { integerValue: 1 } },
+    b: { k: { stringValue: 'b' }, bar: { integerValue: 2 } }
+  };
+
+  function bundleWithTestDocs(
+    db: firestore.FirebaseFirestore
+  ): TestBundleBuilder {
+    const a = key('coll-1/a');
+    const b = key('coll-1/b');
+    const builder = new TestBundleBuilder(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (db as any)._databaseId as DatabaseId
+    );
+    builder.addDocumentMetadata(a, { seconds: 1000, nanos: 9999 }, true);
+    builder.addDocument(
+      a,
+      { seconds: 1, nanos: 9 },
+      { seconds: 1, nanos: 9 },
+      testDocs.a
+    );
+    builder.addDocumentMetadata(b, { seconds: 1000, nanos: 9999 }, true);
+    builder.addDocument(
+      b,
+      { seconds: 1, nanos: 9 },
+      { seconds: 1, nanos: 9 },
+      testDocs.b
+    );
+
+    return builder;
+  }
+
+  it('load with documents only with on progress and promise interface.', () => {
+    return withTestDb(persistence, async db => {
+      const builder = bundleWithTestDocs(db);
+
+      const progresses: firestore.LoadBundleTaskProgress[] = [];
+      let completeProgress: firestore.LoadBundleTaskProgress,
+        fulfillProgress: firestore.LoadBundleTaskProgress;
+      await db
+        .loadBundle(
+          builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+        )
+        .onProgress(
+          progress => {
+            progresses.push(progress);
+          },
+          err => {
+            throw err;
+          },
+          progress => {
+            completeProgress = progress!;
+            return progress;
+          }
+        )
+        .then(progress => {
+          fulfillProgress = progress;
+        })
+        .catch(err => {
+          throw err;
+        });
+
+      verifySuccessProgress(completeProgress!);
+      verifySuccessProgress(fulfillProgress!);
+      expect(progresses.length).to.equal(3);
+      verifyInProgress(progresses[0], 0);
+      verifyInProgress(progresses[1], 1);
+      verifyInProgress(progresses[2], 2);
+
+      // Read from cache. These documents do not exist in backend, so they can
+      // only be read from cache.
+      const snap = await db.collection('coll-1').get({ source: 'cache' });
+      expect(toDataArray(snap)).to.deep.equal([
+        { k: 'a', bar: 1 },
+        { k: 'b', bar: 2 }
+      ]);
```

Comment:
Since the document creation is hidden away in a helper, it might make sense to also move the verification step into a helper. We already cannot really tell where these values come from.